### PR TITLE
Refactor exp.h from reference_ops.h; 2nd step to port op EXP for TFLi…

### PR DESCRIPTION
**Pull request by rsun-bdti**
Issue #45415 PR2

### CODE REVIEW

**Review by petewarden — APPROVED** [review]

File: tensorflow/lite/kernels/internal/reference/exp.h
Comment (petewarden) [inline tensorflow/lite/kernels/internal/reference/exp.h:28]: We need to include the RUY header here.

Comment (rsun-bdti) [reply tensorflow/lite/kernels/internal/reference/exp.h:29]: Added. See line 20.

**Comment by advaitjain** [comment]
This PR is failing internal checks which can be reproduced with:
```
bazel test tensorflow/lite/micro/...  --test_tag_filters=-no_oss --build_tag_filters=-no_oss --copt=-DTF_LITE_STATIC_MEMORY
```

unfortunately, the command above is also failing for unrelated reasons (because we do not have that build variant as part of our CI).

I'm going to fix both of these issues and then you should be able to see what error is being caused by this particular PR.

Please wait for me to prepare and merge a PR to fix the underlying issue -- I'll let you know when that happens.

**Comment by advaitjain** [comment]
> This PR is failing internal checks which can be reproduced with:
> 
> ```
> bazel test tensorflow/lite/micro/...  --test_tag_filters=-no_oss --build_tag_filters=-no_oss --copt=-DTF_LITE_STATIC_MEMORY
> ```
> 
> unfortunately, the command above is also failing for unrelated reasons (because we do not have that build variant as part of our CI).
> 
> I'm going to fix both of these issues and then you should be able to see what error is being caused by this particular PR.
> 
> Please wait for me to prepare and merge a PR to fix the underlying issue -- I'll let you know when that happens.

https://github.com/tensorflow/tensorflow/pull/45679 should do the trick.

**Comment by rsun-bdti** [comment]
Thanks @advaitjain for fixing the failure. Curious to know why the failure was triggered. I assume that refactoring <op>.h is a routine step in porting any TFL op to TFLM, and many ops have been ported by doing that. I am a little surprised that I encountered this failure.

**Comment by advaitjain** [comment]
I should have been more explicit in my comment. My intent with #45679 was not to fix the error being caused by the current PR (which I believe you have done with the ruy header). Rather, it was intended to make the external CI better match the internal CI.

For the current CL, the discrepancy between internal and external fixed with #45679 is a red-herring. What appears to be happening is that when I looked at the change internally (it did not have the ruy header, hence the error) but when I looked at the change on github you had already fixed the missing include (hence it was green here).

Importing of a github pull request into the internal review system is not blocked by a failing external CI, which caused the confusion for me.

**Comment by rsun-bdti** [comment]
@advaitjain Thanks for the explanation. It all makes sense now.

**Review by advaitjain — APPROVED** [review]

**Comment by gbaned** [comment]
@rsun-bdti  Can you please resolve conflicts? Thanks!

**Comment by rsun-bdti** [comment]
I just submitted the merge to resolve the conflicts. However, it was strange that the conflicts occurred. Those lines should have been in the master after my PR1 for issue #45415 was approved and merged.
-Robby

**Comment by rsun-bdti** [comment]
Actually the lines in conflicts were not in my PR1 for issue #45415. They were part of this PR2. I see new lines added into BUILD and reference_ops.h, which my PR2 is also trying to update. Thus the conflicts.
-Robby

From: Robby Sun
Sent: Wednesday, December 16, 2020 8:27 PM
To: tensorflow/tensorflow <reply@reply.github.com>; tensorflow/tensorflow <tensorflow@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [tensorflow/tensorflow] Refactor exp.h from reference_ops.h; 2nd step to port op EXP for TFLi… (#45535)

I just submitted the merge to resolve the conflicts. However, it was strange that the conflicts occurred. Those lines should have been in the master after my PR1 for issue #45415 was approved and merged.
-Robby

**Comment by rsun-bdti** [comment]
@gbaned I merged tensorflow/lite/kernels/internal/reference/reference_ops.h and tensorflow/lite/kernels/internal/BUILD into my branch, but am now faced with two internal CI build failures: Windows Bazel, and Windows Bazel GPU.

**Comment by advaitjain** [comment]
The interal CI build errors are unrelated to the current PR. You too can click through and see the logs (though its likely easier for a Tensorflow engineer to make a decision that the errors are unrelated).

I think that this PR is good to go. I'll let Pete take a final pass an approve.

**Comment by rsun-bdti** [comment]
"Intel® oneDNN -- Community CI Build — oneDNN unit tests failed" 
Executed 1186 out of 1186 tests: 1185 tests pass and 1 fails locally.
//tensorflow/core/kernels:quantize_op_test                               FAILED in 3 out of 3 in 0.7s

The failure doesn't seem to be related to my PR.

**Review by petewarden — APPROVED** [review]

**Pull request merged**
